core/src/main/java/io/questdb/cairo/mv/MatViewRefreshList.java

```diff
@@ -0,0 +1,91 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.mv;
+
+import io.questdb.cairo.TableToken;
+import io.questdb.std.ObjList;
+import io.questdb.std.SimpleReadWriteLock;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+
+/**
+ * Refresh list serves two purposes. First, it holds the list of mat view tokens for the given
+ * base table (or base mat view). The list is protected with a R/W mutex, so it can be read
+ * concurrently. Second, it tracks the last base table txn for which a notification message
+ * was sent to one of {@link MatViewRefreshJob}s. The goal here is to avoid sending excessive
+ * incremental refresh messages to the underlying queue.
+ */
+public class MatViewRefreshList {
+    // Flips to negative value once a refresh message is processed.
+    // Long.MIN_VALUE stands for "just invalidated" state.
+    private final AtomicLong lastNotifiedBaseTableTxn = new AtomicLong(0);
+    private final ReadWriteLock lock = new SimpleReadWriteLock();
+    private final ObjList<TableToken> matViews = new ObjList<>();
+
+    // Called by refresh job once it invalidated dependent mat views.
+    public void notifyOnBaseInvalidated() {
+        lastNotifiedBaseTableTxn.set(Long.MIN_VALUE);
+    }
+
+    // Called by WAL apply job once it applied transaction(s) to the base table
+    // and wants to send refresh job an incremental refresh message.
+    public boolean notifyOnBaseTableCommitNoLock(long seqTxn) {
+        long lastNotified;
+        boolean retry;
+        do {
+            lastNotified = lastNotifiedBaseTableTxn.get();
+            retry = Math.abs(lastNotified) < seqTxn && !lastNotifiedBaseTableTxn.compareAndSet(lastNotified, seqTxn);
+        } while (retry);
+        // The job is allowed to send a notification once incremental refresh finishes.
+        return lastNotified <= 0;
+    }
+
+    // Called by refresh job once it incrementally refreshed dependent mat views.
+    public boolean notifyOnBaseTableRefreshedNoLock(long seqTxn) {
+        // Flip the sign bit in the last notified base table txn number.
+        lastNotifiedBaseTableTxn.compareAndSet(seqTxn, -seqTxn);
+        long lastNotified = lastNotifiedBaseTableTxn.get();
+        return lastNotified != Long.MIN_VALUE && lastNotified != -seqTxn;
+    }
+
+    ObjList<TableToken> lockForRead() {
```

Comment:
nit
```suggestion
    ReadOnlyObjList<TableToken> lockForRead() {
```

Reply:
Changed in 12574dcabc923eae0b7d3f903e30d2c17c7ca945